tck/src/main/java/org/eclipse/microprofile/config/tck/profile/DevConfigProfileTest.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.eclipse.microprofile.config.tck.profile;
+
+
+import static org.hamcrest.MatcherAssert.assertThat;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+
+
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.spi.CDI;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.eclipse.microprofile.config.tck.base.AbstractTest;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.testng.Arquillian;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.Test;
+
+/**
+ * Test cases for Config profile
+ * 
+ * @author Emily Jiang
+ */
+public class DevConfigProfileTest extends Arquillian {
+
+    
+    @Deployment
+    public static Archive deployment() {
+        JavaArchive testJar = ShrinkWrap
+                .create(JavaArchive.class, "DevConfigProfileTest.jar")
+                .addClasses(DevConfigProfileTest.class, ProfilePropertyBean.class)
+                .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml")
+                .as(JavaArchive.class);
+
+        AbstractTest.addFile(testJar, "META-INF/microprofile-config.properties");
+
+        WebArchive war = ShrinkWrap
+                .create(WebArchive.class, "DevConfigProfileTest.war")
+                .addAsLibrary(testJar);
+        return war;
+       
+    }
+
+    @BeforeTest
+    public void setUpTest() {
+        clearAllPropertyValues();
+        ensureAllPropertyValuesAreDefined();
+    }
```

Comment:
We may want to figure out a different way to pass in the profile name to test.

The main issue is that `@BeforeTest` runs before any test being executed and not for this class only. It means that every `@BeforeTest` will execute and provide a non deterministic value for the config. `@BeforeClass` doesn't work either, because it runs after the Arquillian deployment is done (and the Config initialised). 

We can leave it this way, but will have to set up separate `test` with the specific classes of profiles or maybe we can just put the profile name in `microprofile-config.properties` and change it on the fly with the deployment.

Reply:
The profile name can be set via system property or environment variable. We can't set it in microprofile-config.properties unfortunately.

Reply:
Why not?

Reply:
It is because in the spec we said the profile name is set via system property or environment variable. How can we directly set it in a configuration file in the tck?

Reply:
Well a profile should be like any other config property, so I think we can set it in any ConfigSource.

Anyway, as we have now these tests will not pass because they don't get their own value when the test starts (either is set before for all tests or it is too late when the Config is already created).

If we want to keep it this way, then we need to provide a suite that runs these tests independently.

Reply:
@radcortez profile needs to be read before reading any other config sources. If we allow specifying this property in any config sources, we need to gather all config sources and work out the ordinal and then figure out the `profile` value.

We have used this approach for other tests. See [here](https://github.com/eclipse/microprofile-config/blob/master/tck/src/main/java/org/eclipse/microprofile/config/tck/CDIPlainInjectionTest.java#L71).

Reply:
I've tested it with SmallRye and it doesn't work. All methods annotated with `@BeforeTest` execute before any actual class execute. So for instance if you are running the entire suite, you may have values set from other test classes, unless you run each class separately.

Reply:
Thanks @radcortez for bearing with me! I see what you meant now. One possible solution is to create multiple test suite for these profile test. The other suggestion is to set only one value for `mp.config.profile`, which is not ideal.

Reply:
I'll define different test suites, which should fix this issue.

Reply:
hmmm. It is not going well as it needs to launch a different jvm. Here is what I think we can do:

- enable specifying mp.config.profile in microprofile-config.properties
or 

- continue with the current set but test to see which profile is enabled and configure the tests accordingly.
e.g. testConfigProfileWithDev will have a precondition check for mp.config.profile. If it is not set to dev, the tests will be skipped.
Thoughts?

Reply:
It didn't work with different suites? Strange, I think it did work for me.

Anyway, most likely the best approach is to just allow to configure `mp.config.profile` in `microprofile-config.properties` and then use different files in the test.

Reply:
Wait, do you mean that we need to amend the spec to be able to test it? For me this is a no go. This is a significant change as the spec now says that the profile property is only read once during startup. If implementations need to support reading the profile property from a property file in the deployed application, they need to reconfigure config each time an application is deployed. Some implementations provide internal config sources and they would need to reconfigure them too. And I'm not even talking about implementations that support multiple deployments.

If we can't find any better solution, we could create different profiles for tests and require implementers to run the TCK for each profile separately. Some implementations would need to be restarted with a different system property before each TCK.

Alternatively, we can use test suites and provide a hook for implementers to reconfigure/restart their implementation before each suite is run.

Reply:
By the way, a requirement to read the profile from `microprofile-config.properties` again introduces recursion because this config source would need to be initialized before all the other properties can be read. Some implementations (including Payara) initialize some config sources before an application is fully deployed and before config sources in the application are initialized. So I'm really against reading the profile value from any config source.

Reply:
No, there is no limitation to test this. Having different test suits work.

On the other hand, I don't see a reason for the profile configuration to be treated differently to any other configuration property.

Do you read any config before adding the application config sources?

Reply:
> Do you read any config before adding the application config sources?

It really depends. The current wording doesn't specify when the profile value should be read. It only says it's read once and then changes to the profile value are ignored. If implementations read the profile value during app startup but before the application config source is initialized then it's a problem. The spec would need to mandate that the profile value is read only after the microprofile-config.properties source is initialized. If we allow it for all sources, then it needs to be read after all config sources are initialized. 

There are many potential problems for some types of implementations:

* implementations that allow deployment of an application after startup can't guarantee that the profile value is read after the application is deployed
* some implementations allow multiple applications to be deployed. If they both define a different profile, which should be used?
* some implementations themselves are configured by MP Config and read config values during startup, before the app is started and potentially before the microprofile-config.properties source in the app is initialized. So they would ignore the profile defined by the application because it would be too late

For example, Payara allows using MP config values in the server configuration to externalize internal configuration. This configuration is read very early during bootstrap and it's almost impossible to delay it after an app is deployed.

So if the test suite can be adapted to work with a profile defined by a system property or env var, I wouldn't really introduce support for profile definition in MP config sources as it introduces a lot of complexity for very little gain. Of course, implementations can experiment and support it or can allow specifying the profile by other means on top of the specification.

Reply:
> So if the test suite can be adapted to work with a profile defined by a system property or env var, I wouldn't really introduce support for profile definition in MP config sources as it introduces a lot of complexity for very little gain. Of course, implementations can experiment and support it or can allow specifying the profile by other means on top of the specification.

In our implementation, it is simpler to get the profile from MP config sources than it is to read it from the env and sys props separately.  MP Config is already a framework that abstracts the env and sys properties: if it can't be reused for this purpose then there's a design problem.

Reply:
@OndroMih the property `mp.config.profile` is no different from other except it is static, where the others might be dynamic. I think the design makes MP Config much simplier and more usable. I think you can build your config as before. The easier implementation for this is to transform the config value as the config object contains all of the raw values. I think you cannot filter the properties at the config source level.

Reply:
@OndroMih and when do you actually build the Config instance? I believe that the right thing to do might be to specify that the configuration is read when the Config instance is being created.

Now, it does look weird to me that we are working on a Config Spec and we are introducing configuration keys that don't go through our own API :)

Reply:
@OndroMih I think only you argue for this property to be specified as env or system properties. @struberg @jbee both agreed to use normal property to set `mp.config.profile`. I will specify `mp.config.profile` as a normal property.